### ISSUE TITLE
Not respecting auto startup setting

### ISSUE THREAD

**Issue by AlecRust** [mckenfra/postgresql-mac-preferences#7]
Even with the "Automatically Start PostgreSQL Server on Startup" checkbox unchecked, PostgreSQL always starts with my machine and I have to manually stop it using the button:

![screenshot](https://i.imgur.com/PlWkVet.png)

My PostgreSQL version was installed via Homebrew.

**Comment by mckenfra** [comment]
Thanks for raising this. I had noticed the same issue myself, and had been meaning to look into it. I'll do some investigating now.

I can also see in your screenshot there's a problem with some text getting cut-off due to line-wrapping. I'll try to fix that too!

**Comment by AlecRust** [comment]
@mckenfra cheers for your response! Yeah I'm not sure if it's a problem specific to Homebrew PostgreSQL but I'm having difficulty preventing this from starting on startup ([more info](http://superuser.com/questions/244589/prevent-postgresql-from-running-at-startup/308845?noredirect=1#comment1253117_308845)).

Yup there's a text display issue too, as well as some mixing of `PostgresSQL`/`PostgreSQL` in that window :smiley:

**Comment by mckenfra** [comment]
Please could you try the latest release, v1.2?

Download the latest release at: https://github.com/mckenfra/postgresql-mac-preferences/releases

Let me know if that's working now.

**Comment by AlecRust** [comment]
Thanks for this @mckenfra!

- I installed the newer version overwriting the previous. All the display issues fixed etc.
- I "checked" and then "unchecked" the box just to be sure my preference for not launching at startup had been registered.
- After restarting unfortunately PostgreSQL was running and this setting hadn't been respected.
- I stopped PostgreSQL with the button and restarted again, but still it launches on startup.

What mine looks like now:

![screenshot](https://i.imgur.com/Ye7p0dd.png)

**Comment by mckenfra** [comment]
Ok, that's a shame it's still not working! Are you ok to help debug a bit? Can you run the following sets of commands in a Terminal window, then post the results back here?

**1. Check for running PostgreSQL agents in launchctl**

    launchctl list | grep postgre

**2. Check settings in PostgreSQL .plist files**

    find /Library/LaunchDaemons -name "*postgre*" -exec /usr/libexec/Plistbuddy -c print {} \;
    find /Library/LaunchAgents -name "*postgre*" -exec /usr/libexec/Plistbuddy -c print {} \;
    find ~/Library/LaunchAgents -name "*postgre*" -exec /usr/libexec/Plistbuddy -c print {} \;

**3. Check for PostgreSQL in launchctl overrides list***

    my_id=`id -u\; cat /var/db/launchd.db/com.apple.launchd.peruser.${my_id}/overrides.plist

**Comment by AlecRust** [comment]
Sure thing.

**1.**
```sh
(no results)
```

**2.**
Results for last command:
```sh
Dict {
    StandardErrorPath = /usr/local/var/postgres/server.log
    UserName = alec
    EnvironmentVariables = Dict {
        PGDATA = /usr/local/var/postgres
    }
    WorkingDirectory = /usr/local/var/postgres
    Disabled = true
    ProgramArguments = Array {
        /usr/local/opt/postgresql/bin/postgres
        -r
        /usr/local/var/postgres/server.log
    }
    KeepAlive = true
    RunAtLoad = false
    Label = com.hkwebentrepreneurs.postgresql
}
Dict {
    StandardErrorPath = /usr/local/var/postgres/server.log
    WorkingDirectory = /usr/local
    Disabled = true
    ProgramArguments = Array {
        /usr/local/opt/postgresql/bin/postgres
        -D
        /usr/local/var/postgres
        -r
        /usr/local/var/postgres/server.log
    }
    KeepAlive = true
    RunAtLoad = false
    Label = homebrew.mxcl.postgresql
}
```

**3.**
```sh
cat: /var/db/launchd.db/com.apple.launchd.peruser.501/overrides.plist: No such file or directory
```

**Comment by mckenfra** [comment]
So the stuff you posted in (2) looks correct - both of those `.plist` files have `Disabled=true` and `RunAtLoad=false`, which is what we want.

Not sure why (3) doesn't return anything. If you do `ls /var/db/launchd.db` does it return anything? What version of OS X are you running?

But the intriguing one is (1) not returning anything. Can you run this now instead?

    sudo launchctl list | grep postgre

**Comment by AlecRust** [comment]
- `ls /var/db/launchd.db` doesn't return anything.
- I'm running OS X 10.10.3.
- Running `sudo launchctl list | grep postgre` gives me no result!

**Comment by mckenfra** [comment]
Aha, think I've found it. There's a new directory-scheme for launchctl overrides in OS X 10.10 (I've been testing this tool in OS X 10.9).

Just to be sure, can you run this and post the results?

    my_id=`id -u`; cat /var/db/com.apple.xpc.launchd/disabled.${my_id}.plist

If (as I expect) this is the reason it's not working for you, then I can post an update to the tool with a fix.

**Comment by AlecRust** [comment]
That seems to be it! Here's the result:

```sh
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>com.apple.TMHelperAgent.SetupOffer</key>
	<false/>
	<key>com.shazam.mac.ShazamHelper</key>
	<true/>
	<key>com.adobe.AdobeCreativeCloud</key>
	<false/>
	<key>2BUA8C4S2C.com.agilebits.onepassword4-helper</key>
	<false/>
	<key>com.apple.rcd</key>
	<true/>
	<key>com.apple.photostream-agent</key>
	<true/>
	<key>com.apple.FileStatsAgent</key>
	<true/>
	<key>com.hkwebentrepreneurs.postgresql</key>
	<false/>
	<key>com.datadoghq.Agent</key>
	<true/>
	<key>com.adobe.AAM.Scheduler-1.0</key>
	<false/>
	<key>com.apple.mrt.uiagent</key>
	<false/>
	<key>com.spotify.webhelper</key>
	<false/>
</dict>
</plist>
```

**Comment by mckenfra** [comment]
Ok, I've just published release v1.3, which should fix the issue. Can you give it a try and let me know?

Download the latest release at: https://github.com/mckenfra/postgresql-mac-preferences/releases

**Comment by AlecRust** [comment]
Upgraded, no dice I'm afraid :(

After restart:
![screenshot](https://i.imgur.com/8ADAt3M.png)

Btw my machine is now running OS X 10.10.4.

**Comment by mckenfra** [comment]
You need to click the checkbox once then uncheck it again, before
restarting. That way, it should remove the override. Did you try that?

**Comment by AlecRust** [comment]
Yup, can confirm I've tried this.

**Comment by mckenfra** [comment]
Ok thanks - last thing, can you run this command again and post the output?

    my_id=`id -u`; cat /var/db/com.apple.xpc.launchd/disabled.${my_id}.plist

In the meantime, I'll do some testing of the tool on OS X 10.10 - must be some other issue. Frustrating!

**Comment by AlecRust** [comment]
Sure:

```sh
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>com.apple.TMHelperAgent.SetupOffer</key>
	<false/>
	<key>com.shazam.mac.ShazamHelper</key>
	<true/>
	<key>com.apple.macbuddy.icloudsetup.user</key>
	<false/>
	<key>com.adobe.AdobeCreativeCloud</key>
	<false/>
	<key>2BUA8C4S2C.com.agilebits.onepassword4-helper</key>
	<false/>
	<key>com.apple.rcd</key>
	<true/>
	<key>com.apple.photostream-agent</key>
	<true/>
	<key>com.hkwebentrepreneurs.postgresql</key>
	<false/>
	<key>com.apple.FileStatsAgent</key>
	<true/>
	<key>com.datadoghq.Agent</key>
	<true/>
	<key>com.adobe.AAM.Scheduler-1.0</key>
	<false/>
	<key>com.apple.mrt.uiagent</key>
	<false/>
	<key>com.spotify.webhelper</key>
	<false/>
</dict>
</plist>
```

**Comment by mckenfra** [comment]
Strange, I just did a test on my OS X 10.10 system, and it correctly removed the override that is causing the problem, whereas I can see the the override for `com.hkwebentrepreneurs.postgresql` is still present in your override file.

Can we just check you've got the right version of the pref pane installed? If you installed it "for this user only", run this command:

    /usr/libexec/PlistBuddy ~/Library/PreferencePanes/PostgreSQL.prefPane/Contents/Info.plist -c Print:CFBundleVersion

If you installed it "for all users of this computer", run this command:

    /usr/libexec/PlistBuddy /Library/PreferencePanes/PostgreSQL.prefPane/Contents/Info.plist -c Print:CFBundleVersion

Can you post up the result?

**Comment by AlecRust** [comment]
`1.3` is the output :smiley:

**Comment by mckenfra** [comment]
Very strange. Can we try running the underlying script in debug mode, and then you post up here the messages that get written to the log file?

Run this command (assuming you installed the prefpane "for this user only"):

    ~/Library/PreferencePanes/PostgreSQL.prefPane/Contents/Resources/PGPrefsPostgreSQL.sh --PGBIN=/usr/local/opt/postgresql/bin/ --PGDATA=/usr/local/var/postgres --DEBUG=Yes autooff

Then run this command to print out the log file contents, and post that up here:

    cat ~/Library/Logs/PostgreSQL/Helpers.log
 
Finally, can you also post up the results of this command, which checks if the key still exists in the override file:

    my_id=`id -u`; sudo /usr/libexec/Plistbuddy /var/db/com.apple.xpc.launchd/disabled.${my_id}.plist -c Print:com.hkwebentrepreneurs.postgresql

**Comment by AlecRust** [comment]
@mckenfra I installed for all users actually, `1.3` was the output of the second command in your previous comment.

Running the command without the `~`:
```sh
/Library/PreferencePanes/PostgreSQL.prefPane/Contents/Resources/PGPrefsPostgreSQL.sh --PGBIN=/usr/local/opt/postgresql/bin/ --PGDATA=/usr/local/var/postgres --DEBUG=Yes autooff
```
Strangely asks me for my password three times then gives no other output.

Here's the result of running `cat ~/Library/Logs/PostgreSQL/Helpers.log` after that:
```sh
Thu 2 Jul 2015 16:30:39 BST

---------------------------------------------------------------------------------
[2015-07-02 16:30:42 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Checking command line args...

PGPrefsPostgreSQL.sh --PGBIN=/usr/local/opt/postgresql/bin/ --PGDATA=/usr/local/var/postgres --DEBUG=Yes autooff

---------------------------------------------------------------------------------
[2015-07-02 16:30:42 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Checking environment variables...

REALUSER | alec
RUNUSER  | alec
PGBIN    | /usr/local/opt/postgresql/bin/
PGDATA   | /usr/local/var/postgres
PGUSER   |
PGLOG    |
PGPORT   |
PGAUTO   |

---------------------------------------------------------------------------------
[2015-07-02 16:30:42 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Checking for launchctl overrides for com.hkwebentrepreneurs.postgresql in /var/db/com.apple.xpc.launchd ...

---------------------------------------------------------------------------------
[2015-07-02 16:30:42 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Running command...

sudo find /var/db/com.apple.xpc.launchd -name "disabled.*.plist" -exec /usr/libexec/Plistbuddy {} -c Print:com.hkwebentrepreneurs.postgresql \; 2>/dev/null | grep -v "Error"

---------------------------------------------------------------------------------
[2015-07-02 16:30:44 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

false

[OK]

---------------------------------------------------------------------------------
[2015-07-02 16:30:44 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Removing launchctl overrides for com.hkwebentrepreneurs.postgresql in /var/db/com.apple.xpc.launchd ...

---------------------------------------------------------------------------------
[2015-07-02 16:30:44 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Running command...

sudo find /var/db/com.apple.xpc.launchd -name "disabled.*.plist" -exec /usr/libexec/Plistbuddy {} -c Delete:com.hkwebentrepreneurs.postgresql \; 2>/dev/null | grep -v "Error"

---------------------------------------------------------------------------------
[2015-07-02 16:30:44 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------



[FATAL]

---------------------------------------------------------------------------------
[2015-07-02 16:30:44 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Running command...

sudo find /var/db/com.apple.xpc.launchd -name "disabled.*.plist" -exec /usr/libexec/Plistbuddy {} -c Print:com.hkwebentrepreneurs.postgresql \; 2>/dev/null | grep -v "Error"

---------------------------------------------------------------------------------
[2015-07-02 16:30:44 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------



[FATAL]
```

The result of running:
```sh
my_id=`id -u`; sudo /usr/libexec/Plistbuddy /var/db/com.apple.xpc.launchd/disabled.${my_id}.plist -c Print:com.hkwebentrepreneurs.postgresql
```
is:
```sh
Print: Entry, ":com.hkwebentrepreneurs.postgresql", Does Not Exist
```

**Comment by mckenfra** [comment]
Yes, it asks for your password because it runs some commands as root. In the preference pane itself, this is handled by requiring the user to click the lock to authenticate first.

So, from the output of the final command it looks like the `com.hkwebentrepreneurs.postgresql` override has now been removed. So from now on, the check-box setting in the preference pane should be honoured, i.e. auto-startup-at-login or otherwise should work as expected. Do you mind giving it another try?

The frustrating thing is, assuming it does work, I'm no clearer as to why when you used the preference pane itself, it didn't remove the override, since it seemed to work fine when we ran exactly the same script from the command line...

Anyway, let me know if it's now fixed.

**Comment by AlecRust** [comment]
I'm afraid PostgreSQL still starts on system boot, I have to manually stop it. Did check and uncheck the checkbox before restarting.

**Comment by mckenfra** [comment]
And if you now run this command again? Override is still gone?

    my_id=`id -u`; sudo /usr/libexec/Plistbuddy /var/db/com.apple.xpc.launchd/disabled.${my_id}.plist -c Print:com.hkwebentrepreneurs.postgresql

**Comment by AlecRust** [comment]
Returns `false`. I did manually stop PostgreSQL before this after my last comment.

**Comment by mckenfra** [comment]
Ok, so if it returns `false` that means the override exists. If it had returned `Print: Entry, ":com.hkwebentrepreneurs.postgresql", Does Not Exist` that would have meant the override did not exist, like after we ran the script manually. We don't want the override to exist!

My hunch is that for some reason the PostgreSQL preference pane in your System Preference is still using the old version of the script. Possibly due to Mac OS X caching the script somewhere?

So can you do this:

1. Right-click on the preference pane icon and click `Remove "PostgreSQL" Preference Pane`
2. Quit System Preferences, just to be safe.
3. Reinstall **version 1.3** of the preference pane.
4. In the PostgreSQL preference pane, click the checkbox on and off once
5. Finally, run this command again:

    my_id=`id -u`; sudo /usr/libexec/Plistbuddy /var/db/com.apple.xpc.launchd/disabled.${my_id}.plist -c Print:com.hkwebentrepreneurs.postgresql

If that now returns `Print: Entry, ":com.hkwebentrepreneurs.postgresql", Does Not Exist` instead of `false` then we may finally have solved it.

If not, I can only think we could try looking in `~/Library/Caches` for the `PGPrefsPostgreSQL.sh` script, and removing it.

Let me know how it goes!

**Comment by AlecRust** [comment]
I followed those steps and restarted, but PostgreSQL still started at boot.

Worth noting that the PostgreSQL Server status was "Stopped" before I checked the checkbox to "ticked". When I checked it the status changed to "Running". When I then unchecked the box the status remained at "Running". I would have thought this checkbox wouldn't interfere with the current status?

The output of that command is still `false`.

I had a look within `~/Library/Caches` but couldn't find anything related to PostgreSQL.

**Comment by mckenfra** [comment]
I just can't understand it. Ok, this is the last thing I can think of. Can you do this:

1) Edit `/Library/PreferencePanes/PostgreSQL.prefPane/Contents/Resources/PGPrefsPostgreSQL.sh`
2) Change this line:

    --DEBUG)     export DEBUG=${lval};;

to:

    --DEBUG)     export DEBUG="yes";;

3) Delete this file (if it exists): `~/Library/Logs/PostgreSQL/Helpers.log`
4) Quit System Preferences (if it's running)
5) Open System Preferences and go into the PostgreSQL preference pane
6) Check and then uncheck the Auto-Startup checkbox once

Then post up here everything that's contained in `~/Library/Logs/PostgreSQL/Helpers.log`

**Comment by AlecRust** [comment]
After those steps, this is the output:

```sh
Thu Jul 2 23:37:26 BST 2015

---------------------------------------------------------------------------------
[2015-07-02 23:37:26 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Checking command line args...

PGPrefsPostgreSQL.sh --PGDATA=/usr/local/var/postgres --PGBIN=/usr/local/opt/postgresql/bin --PGLOG=/usr/local/var/postgres/server.log --PGAUTO=No --DEBUG=No status

---------------------------------------------------------------------------------
[2015-07-02 23:37:26 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Checking environment variables...

REALUSER | alec
RUNUSER  | root
PGBIN    | /usr/local/opt/postgresql/bin
PGDATA   | /usr/local/var/postgres
PGUSER   | 
PGLOG    | /usr/local/var/postgres/server.log
PGPORT   | 
PGAUTO   | no

---------------------------------------------------------------------------------
[2015-07-02 23:37:26 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Running command...

su alec -c "export PGDATA="/usr/local/var/postgres"; "/usr/local/opt/postgresql/bin/pg_ctl" -l "/usr/local/var/postgres/server.log" -m fast status"

---------------------------------------------------------------------------------
[2015-07-02 23:37:26 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

pg_ctl: no server running

[FATAL]

---------------------------------------------------------------------------------
[2015-07-02 23:37:37 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Checking command line args...

PGPrefsPostgreSQL.sh --PGDATA=/usr/local/var/postgres --PGBIN=/usr/local/opt/postgresql/bin --PGLOG=/usr/local/var/postgres/server.log --PGAUTO=Yes --DEBUG=No AutoOn

---------------------------------------------------------------------------------
[2015-07-02 23:37:37 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Checking environment variables...

REALUSER | alec
RUNUSER  | root
PGBIN    | /usr/local/opt/postgresql/bin
PGDATA   | /usr/local/var/postgres
PGUSER   | 
PGLOG    | /usr/local/var/postgres/server.log
PGPORT   | 
PGAUTO   | yes

---------------------------------------------------------------------------------
[2015-07-02 23:37:37 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Checking for launchctl overrides for com.hkwebentrepreneurs.postgresql in /var/db/com.apple.xpc.launchd ...

---------------------------------------------------------------------------------
[2015-07-02 23:37:37 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Running command...

sudo find /var/db/com.apple.xpc.launchd -name "disabled.*.plist" -exec /usr/libexec/Plistbuddy {} -c Print:com.hkwebentrepreneurs.postgresql \; 2>/dev/null | grep -v "Error"

---------------------------------------------------------------------------------
[2015-07-02 23:37:37 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

false

[OK]

---------------------------------------------------------------------------------
[2015-07-02 23:37:37 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Removing launchctl overrides for com.hkwebentrepreneurs.postgresql in /var/db/com.apple.xpc.launchd ...

---------------------------------------------------------------------------------
[2015-07-02 23:37:37 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Running command...

sudo find /var/db/com.apple.xpc.launchd -name "disabled.*.plist" -exec /usr/libexec/Plistbuddy {} -c Delete:com.hkwebentrepreneurs.postgresql \; 2>/dev/null | grep -v "Error"

---------------------------------------------------------------------------------
[2015-07-02 23:37:37 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------



[FATAL]

---------------------------------------------------------------------------------
[2015-07-02 23:37:37 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Running command...

sudo find /var/db/com.apple.xpc.launchd -name "disabled.*.plist" -exec /usr/libexec/Plistbuddy {} -c Print:com.hkwebentrepreneurs.postgresql \; 2>/dev/null | grep -v "Error"

---------------------------------------------------------------------------------
[2015-07-02 23:37:37 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------



[FATAL]

---------------------------------------------------------------------------------
[2015-07-02 23:37:37 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Unloading launchagents for root ...

---------------------------------------------------------------------------------
[2015-07-02 23:37:37 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Running command...

sudo launchctl list | grep postgresql | cut -f 3

---------------------------------------------------------------------------------
[2015-07-02 23:37:37 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------



[OK]

---------------------------------------------------------------------------------
[2015-07-02 23:37:37 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Unloading launchagents for alec ...

---------------------------------------------------------------------------------
[2015-07-02 23:37:37 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Running command...

su alec -c "launchctl list | grep postgresql | cut -f 3"

---------------------------------------------------------------------------------
[2015-07-02 23:37:37 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------



[OK]

---------------------------------------------------------------------------------
[2015-07-02 23:37:37 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Already Running: 2613  2476 root            /usr/bin/osascript /Library/PreferencePanes/PostgreSQL.prefPane/Contents/Resources/PGPrefsRunAsAdmin.scpt /Library/PreferencePanes/PostgreSQL.prefPane/Contents/Resources/PGPrefsPostgreSQL.sh "--PGDATA=/usr/local/var/postgres" "--PGBIN=/usr/local/opt/postgresql/bin" "--PGLOG=/usr/local/var/postgres/server.log" --PGAUTO=Yes --DEBUG=No AutoOn

---------------------------------------------------------------------------------
[2015-07-02 23:37:37 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Already Running: 2614  2613 root            /bin/sh /Library/PreferencePanes/PostgreSQL.prefPane/Contents/Resources/PGPrefsPostgreSQL.sh --PGDATA=/usr/local/var/postgres --PGBIN=/usr/local/opt/postgresql/bin --PGLOG=/usr/local/var/postgres/server.log --PGAUTO=Yes --DEBUG=No AutoOn

---------------------------------------------------------------------------------
[2015-07-02 23:37:37 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Already Running: 2893  2614 root            /bin/sh /Library/PreferencePanes/PostgreSQL.prefPane/Contents/Resources/PGPrefsPostgreSQL.sh --PGDATA=/usr/local/var/postgres --PGBIN=/usr/local/opt/postgresql/bin --PGLOG=/usr/local/var/postgres/server.log --PGAUTO=Yes --DEBUG=No AutoOn

---------------------------------------------------------------------------------
[2015-07-02 23:37:38 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Deleting postgresql agent plist file(s)...

/Users/alec/Library/LaunchAgents/com.hkwebentrepreneurs.postgresql.plist

---------------------------------------------------------------------------------
[2015-07-02 23:37:38 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Running command...

su alec -c "find /Users/alec/Library/LaunchAgents -maxdepth 1 -name "com.hkwebentrepreneurs.postgresql*.plist" -exec rm {} \;"

---------------------------------------------------------------------------------
[2015-07-02 23:37:38 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------



[OK]

---------------------------------------------------------------------------------
[2015-07-02 23:37:38 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Creating /Users/alec/Library/LaunchAgents/com.hkwebentrepreneurs.postgresql.plist ...

---------------------------------------------------------------------------------
[2015-07-02 23:37:38 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Running command...

su alec -c "echo "<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>Label</key>
  <string>com.hkwebentrepreneurs.postgresql</string>
  <key>EnvironmentVariables</key>
  <dict>
    <key>PGDATA</key>
    <string>/usr/local/var/postgres</string>
  </dict>
  <key>ProgramArguments</key>
  <array>
    <string>/usr/local/opt/postgresql/bin/postgres</string>
    <string>-r</string>
    <string>/usr/local/var/postgres/server.log</string>
  </array>
  <key>UserName</key>
  <string>alec</string>
  <key>WorkingDirectory</key>
  <string>/usr/local/var/postgres</string>
  <key>StandardErrorPath</key>
  <string>/usr/local/var/postgres/server.log</string>
  <key>Disabled</key>
  <false/>
  <key>RunAtLoad</key>
  <true/>
  <key>KeepAlive</key>
  <true/>
</dict>
</plist>" > "/Users/alec/Library/LaunchAgents/com.hkwebentrepreneurs.postgresql.plist""

---------------------------------------------------------------------------------
[2015-07-02 23:37:38 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------



[OK]

---------------------------------------------------------------------------------
[2015-07-02 23:37:38 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Loading /Users/alec/Library/LaunchAgents/com.hkwebentrepreneurs.postgresql.plist ...

---------------------------------------------------------------------------------
[2015-07-02 23:37:38 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Running command...

su alec -c "launchctl load -F /Users/alec/Library/LaunchAgents/com.hkwebentrepreneurs.postgresql.plist"

---------------------------------------------------------------------------------
[2015-07-02 23:37:38 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------



[OK]

---------------------------------------------------------------------------------
[2015-07-02 23:37:38 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Running command...

su alec -c "launchctl list | cut -f 3 | grep ^com.hkwebentrepreneurs.postgresql$"

---------------------------------------------------------------------------------
[2015-07-02 23:37:38 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

com.hkwebentrepreneurs.postgresql

[OK]

---------------------------------------------------------------------------------
[2015-07-02 23:37:41 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Checking command line args...

PGPrefsPostgreSQL.sh --PGDATA=/usr/local/var/postgres --PGBIN=/usr/local/opt/postgresql/bin --PGLOG=/usr/local/var/postgres/server.log --PGAUTO=Yes --DEBUG=No status

---------------------------------------------------------------------------------
[2015-07-02 23:37:41 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Checking environment variables...

REALUSER | alec
RUNUSER  | root
PGBIN    | /usr/local/opt/postgresql/bin
PGDATA   | /usr/local/var/postgres
PGUSER   | 
PGLOG    | /usr/local/var/postgres/server.log
PGPORT   | 
PGAUTO   | yes

---------------------------------------------------------------------------------
[2015-07-02 23:37:41 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Running command...

su alec -c "export PGDATA="/usr/local/var/postgres"; "/usr/local/opt/postgresql/bin/pg_ctl" -l "/usr/local/var/postgres/server.log" -m fast status"

---------------------------------------------------------------------------------
[2015-07-02 23:37:41 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

pg_ctl: server is running (PID: 3052)
/usr/local/Cellar/postgresql/9.4.4/bin/postgres "-r" "/usr/local/var/postgres/server.log"

[OK]

---------------------------------------------------------------------------------
[2015-07-02 23:37:46 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Checking command line args...

PGPrefsPostgreSQL.sh --PGDATA=/usr/local/var/postgres --PGBIN=/usr/local/opt/postgresql/bin --PGLOG=/usr/local/var/postgres/server.log --PGAUTO=No --DEBUG=No AutoOff

---------------------------------------------------------------------------------
[2015-07-02 23:37:46 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Checking environment variables...

REALUSER | alec
RUNUSER  | root
PGBIN    | /usr/local/opt/postgresql/bin
PGDATA   | /usr/local/var/postgres
PGUSER   | 
PGLOG    | /usr/local/var/postgres/server.log
PGPORT   | 
PGAUTO   | no

---------------------------------------------------------------------------------
[2015-07-02 23:37:46 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Running command...

su alec -c "/usr/libexec/PlistBuddy /Users/alec/Library/LaunchAgents/com.hkwebentrepreneurs.postgresql.plist -c "Set:RunAtLoad false""

---------------------------------------------------------------------------------
[2015-07-02 23:37:46 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------



[OK]

---------------------------------------------------------------------------------
[2015-07-02 23:37:46 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Running command...

su alec -c "/usr/libexec/PlistBuddy /Users/alec/Library/LaunchAgents/com.hkwebentrepreneurs.postgresql.plist -c "Set:Disabled true""

---------------------------------------------------------------------------------
[2015-07-02 23:37:46 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------



[OK]

---------------------------------------------------------------------------------
[2015-07-02 23:37:46 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Checking for launchctl overrides for com.hkwebentrepreneurs.postgresql in /var/db/com.apple.xpc.launchd ...

---------------------------------------------------------------------------------
[2015-07-02 23:37:46 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Running command...

sudo find /var/db/com.apple.xpc.launchd -name "disabled.*.plist" -exec /usr/libexec/Plistbuddy {} -c Print:com.hkwebentrepreneurs.postgresql \; 2>/dev/null | grep -v "Error"

---------------------------------------------------------------------------------
[2015-07-02 23:37:47 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------



[FATAL]

---------------------------------------------------------------------------------
[2015-07-02 23:37:50 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Checking command line args...

PGPrefsPostgreSQL.sh --PGDATA=/usr/local/var/postgres --PGBIN=/usr/local/opt/postgresql/bin --PGLOG=/usr/local/var/postgres/server.log --PGAUTO=No --DEBUG=No status

---------------------------------------------------------------------------------
[2015-07-02 23:37:50 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Checking environment variables...

REALUSER | alec
RUNUSER  | root
PGBIN    | /usr/local/opt/postgresql/bin
PGDATA   | /usr/local/var/postgres
PGUSER   | 
PGLOG    | /usr/local/var/postgres/server.log
PGPORT   | 
PGAUTO   | no

---------------------------------------------------------------------------------
[2015-07-02 23:37:50 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

Running command...

su alec -c "export PGDATA="/usr/local/var/postgres"; "/usr/local/opt/postgresql/bin/pg_ctl" -l "/usr/local/var/postgres/server.log" -m fast status"

---------------------------------------------------------------------------------
[2015-07-02 23:37:50 PGPrefsPostgreSQL.sh]
---------------------------------------------------------------------------------

pg_ctl: server is running (PID: 3052)
/usr/local/Cellar/postgresql/9.4.4/bin/postgres "-r" "/usr/local/var/postgres/server.log"

[OK]
```

**Comment by mckenfra** [comment]
Thanks v. much for posting this. The good news is, I've recreated the problem on my OS X 10.10 system. The new version of the preference pane manually removes the `com.hkwebentrepreneurs.postgresql` override, which causes postgresql always to start on login. This override was incorrectly inserted by the older version of the preference pane.

So the new version manually removes it. But then, for some reason, Mac OS X puts the override back in when you restart your computer! So, obviously manually deleting the override isn't the solution. I'll now have to see what other options there are.

Anyway, thanks again for your help in getting us this far! I'll try to get back to you soon with a fix.

**Comment by AlecRust** [comment]
No problem :+1: thanks for your efforts on this!

**Comment by mckenfra** [comment]
Please could you try the latest release, v1.4?

Download the latest release at: https://github.com/mckenfra/postgresql-mac-preferences/releases

Let me know if that's working now.

**Comment by AlecRust** [comment]
@mckenfra that did it! After installing the latest release, checking/unchecking then restarting PostgreSQL Server is "Stopped":

![screenshot](https://i.imgur.com/q182mcO.png)

Thank you again!

**Comment by mckenfra** [comment]
That's great, yes I finally decided that on OS X 10.10 there is no option but to use the permanent overrides feature, so the tool now always sets the override when you click the Auto Startup checkbox.

I raised a stackoverflow question about how to delete the override http://stackoverflow.com/questions/31206756/launchctl-remove-enabled-disabled-override but I'm not expecting any answers, because it doesn't seem possible!

Anyway, thanks for raising the issue and for helping debug it! :+1:
